src/ray/core_worker/core_worker.cc

```diff
@@ -61,6 +62,15 @@ std::unique_ptr<CoreWorkerProcess> CoreWorkerProcess::instance_;
 
 thread_local std::weak_ptr<CoreWorker> CoreWorkerProcess::current_core_worker_;
 
+boost::asio::io_service CoreWorkerProcess::stats_io_service_;
+
+boost::asio::io_service::work CoreWorkerProcess::stats_io_work_(
+    CoreWorkerProcess::stats_io_service_);
+
+std::shared_ptr<std::thread> CoreWorkerProcess::stats_thread_;
+
+absl::Mutex CoreWorkerProcess::stats_mutex_;
```

Comment:
Do we need to have this when `.Record` is called by each metric? Is openCensus Record method thread-safe?

Reply:
> Do we need to have this when `.Record` is called by each metric? Is openCensus Record method thread-safe?

No. It's just used for making  sure stats module will be created exactly once.

Reply:
I can't see how you enforce one-time initialization for stats module. In Python test, almost every test case will invoke `CoreWorkerProcess::Initialzie` and then `CoreWorkerProcess::Shutdown` once. If we run multiple Python test cases, which is a typical scenario, `CoreWorkerProcess::Initialzie` will be invoked multiple times, then `CoreWorkerProcess::RunStatsService()` multiple times too. We should check a flag or something which marks that the stats thread is already running.